### PR TITLE
fix(resources): Incorrect resource resolution for multiple identical source dictionaries

### PR DESCRIPTION
**Pull request by jeromelaban**
GitHub Issue (If applicable): closes https://github.com/unoplatform/uno/issues/13799

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the new behavior?

Ensures that lazy initializers resolve resources using the appropriate XAML name scope.

### <samp>🤖 Generated by Copilot at f58be7b</samp>

This pull request adds a new XAML page and a unit test to verify the behavior of theme resources when they are overridden by named resource dictionaries in different scopes. It also fixes a bug in the `ResourceDictionary.CopyFrom` method that caused lazy resource initializers to use the wrong XAML scope. The changes affect the files `ThemeResource_Named_ResourceDictionary_Override.xaml`, `ThemeResource_Named_ResourceDictionary_Override.xaml.cs`, `Given_ResourceDictionary.cs`, and `ResourceDictionary.cs`.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 1]
```diff
@@ -427,7 +427,23 @@ internal void CopyFrom(ResourceDictionary source)
 			// In order to ensure the theme updates covers all of them, we need to keep track of this source instance for theme updates before it is lost.
 			_sourceDictionary = WeakReferencePool.RentWeakReference(this, source);
 
-			_values.AddRange(source._values);
+			// Lazy resource initialization needs the current XamlScope
+			// to resolve values, and the originally defined scope that 
+			// was set when the source dictionary was created may not be 
+			// value for the current XAML scope. We rewrite the initializer
+			// in order for the name resolution to work properly.
+			foreach (var pair in source._values)
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:435]: Consider calling _values.EnsureCapacity to make sure the repeated Add calls can't cause multiple resizes. This is something was done when using AddRange.

Not very important though, but makes sure we are not regressing performance in this code path

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:432]: Good catch! Thanks!

**Review by kazo0 — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/ResourceResolver.cs [hunk 2]
```diff
@@ -405,14 +405,17 @@ private static bool TryAssemblyResourceRetrieval(in SpecializedResourceDictionar
...
-				foreach (var kvp in assemblyDict)
+				if (_registeredDictionariesByAssembly.TryGetValue(parseContext.AssemblyName, out var assemblyDict))
 				{
-					var rd = kvp.Value as ResourceDictionary;
-					if (rd.TryGetValue(resourceKey, out value, shouldCheckSystem: false))
+					foreach (var kvp in assemblyDict)
 					{
-						return true;
+						var rd = kvp.Value as ResourceDictionary;
+						if (rd.TryGetValue(resourceKey, out value, shouldCheckSystem: false))
```
Comment (ajpinedam) [inline src/Uno.UI/UI/Xaml/ResourceResolver.cs:415]: In this scenario where we don't care about the `out value` isn't more convenient to use `ContainsKey` method?

**Review by ajpinedam — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/ResourceResolver.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceResolver.cs:415]: I think we care about the out value. That is reflected in the out parameter of this method

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ResourceResolver.cs:415]: Indeed, the out value is used as the out value of the containing method.

Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/ResourceResolver.cs:415]: I see it, thanks, I missed it the first time. :)

**Pull request merged**